### PR TITLE
Make vkFrameBoundaryANDROID trigger screenshots

### PR DESCRIPTION
**Pull request by marius-pelegrin-arm**
This commit is part of what is requested in #1494

### CODE REVIEW

**Comment by bradgrantham-lunarg** [comment]
@panos-lunarg @locke-lunarg could you take a look at this PR, please?  It would be helpful to merge for other sponsors so we should prioritize it.  Thank you!

**Comment by locke-lunarg** [comment]
These code is duplicated with `VulkanReplayConsumerBase::WriteScreenshots`. It will be better if it has a function for these two cases.

**Comment by panos-lunarg** [comment]
Some thoughts on `vkFrameBoundaryANDROID` and virtual swapchains:
`VulkanReplayConsumerBase::WriteScreenshots` takes the image information from the `SwapchainKHRInfo` which should take care of virtual swapchains. 
On the other hand `vkFrameBoundaryANDROID` takes as a direct argument a handle to the `VkImage` that's going to be presented/swapped. I assume that the image handles are acquired through `vkGetSwapchainImagesKHR` which takes care of virtual swapchains. So it should be fine.
@locke-lunarg : does this sound correct to you?

**Comment by locke-lunarg** [comment]
I'm not sure if the image on `vkFrameBoundaryANDROID` is from `vkGetSwapchainImagesKHR`. My option is quit simple that it could have a function, like  `VulkanReplayConsumerBase::WriteScreenshots(ImageInfo* image_info)`.

And then
```
VulkanReplayConsumerBase::WriteScreenshots(const Decoded_VkPresentInfoKHR* meta_info)
{
   ...
          WriteScreenshots(image_info);
   ...
}
```

```
VulkanReplayConsumerBase::OverrideFrameBoundaryANDROID(....)
{
   ...
          WriteScreenshots(image_info);
   ...
}
```

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 1]
```diff
@@ -7229,6 +7229,60 @@ VkResult VulkanReplayConsumerBase::OverrideCreateFramebuffer(
...
+                    : (options_.screenshot_height ? options_.screenshot_height : image_info->extent.height);
+
+            auto instance_table = GetInstanceTable(device_info->parent);
+            GFXRECON_ASSERT(instance_table != nullptr);
+
+            VkPhysicalDeviceMemoryProperties memory_properties;
+            instance_table->GetPhysicalDeviceMemoryProperties(device_info->parent, &memory_properties);
+
+            screenshot_handler_->WriteImage(filename_prefix,
+                                            device,
```
Comment (panos-lunarg) [inline framework/decode/vulkan_replay_consumer_base.cpp:7267]: `devivce_info` instead of `device`

**Comment by panos-lunarg** [comment]
> I'm not sure if the image on `vkFrameBoundaryANDROID` is from `vkGetSwapchainImagesKHR`. My option is quit simple that it could have a function, like `VulkanReplayConsumerBase::WriteScreenshots(ImageInfo* image_info)`.

Yes, I agree with this suggestion.
Basically my question was about whether this change is compatible with virtual swapchain. As far as I can tell I think it is.

**Comment by locke-lunarg** [comment]
> Yes, I agree with this suggestion. Basically my question was about whether this change is compatible with virtual swapchain. As far as I can tell I think it is.

Good point. For virtual swapchain, GetSwapchainImagesKHR return virtual swapchain images, not real swapchain images, so `vkFrameBoundaryANDROID` would use virtual swapchain images to output image, instead of real swapchain images. But virtual swapchain images and real swapchain images are the same, so it's not a problem.

File: framework/decode/vulkan_replay_consumer_base.cpp (hunk 1 above)
Comment (marius-pelegrin-arm) [reply framework/decode/vulkan_replay_consumer_base.cpp:7267]: Not sure to understand... WriteImage takes a VkDevice as an argument, not a DeviceInfo

**Comment by marius-pelegrin-arm** [comment]
> I'm not sure if the image on `vkFrameBoundaryANDROID` is from `vkGetSwapchainImagesKHR`. My option is quit simple that it could have a function, like `VulkanReplayConsumerBase::WriteScreenshots(ImageInfo* image_info)`.
> 
> And then
> 
> ```
> VulkanReplayConsumerBase::WriteScreenshots(const Decoded_VkPresentInfoKHR* meta_info)
> {
>    ...
>           WriteScreenshots(image_info);
>    ...
> }
> ```
> 
> ```
> VulkanReplayConsumerBase::OverrideFrameBoundaryANDROID(....)
> {
>    ...
>           WriteScreenshots(image_info);
>    ...
> }
> ```
> 
> to avoid the duplicated code.

Sure the duplication is ugly, but the only code "really" duplicated here is the computation of the screenshot size. I cannot create the proposed function simply because there is no `image_info` accessible from the swapchain, only a vector of `VkImage`.
All the code that could be factorized was factorized in `WriteImage`

File: framework/decode/vulkan_replay_consumer_base.cpp (hunk 1 above)
Comment (panos-lunarg) [reply framework/decode/vulkan_replay_consumer_base.cpp:7267]: Right, I saw a compilation error when I cherry picked this patch in dev. In this branch this is correct but from what I see in current dev, `WriteImage` has changed.

**Review by locke-lunarg — APPROVED** [review]
Thank you. LGTM

**Comment by locke-lunarg** [comment]
Sorry. That's a wrong click.

**Comment by locke-lunarg** [comment]
@marius-pelegrin-arm  Please solve the conflict, and then it's good to merge.

**Comment by bradgrantham-lunarg** [comment]
> Hi, @locke-lunarg ! I resolved a conflict but that wasn't sufficient; there are compile errors. Will you check out Marius' branch and see if the fix is obvious?

Never mind, I think I see the issue.

**Pull request merged**